draft-johani-dnsop-multi-provider-requirements-00.md

```diff
@@ -148,6 +148,199 @@ to be able to fully support all multi-provider scenarios:
+roll is complete.
+
+For the Pre-Publish ZSK roll, provider A first adds ZSK_An to its DNSKEY RRset
+and informs provider B that is now has two ZSKs: ZSK_A and ZSK_An.
+Provider A monitors the propagation of its own zone and waits for confirmation
+from provider B that the new DNSKEY RRset has fully propagated.
+
+Then provider A starts signing with ZSK_An and monitors propagation. This is
+a local operation.
+
+Finally, provider A removes ZSK_A from the DNSKEY RRset and informs provider B
+that it should only include ZSK_An in its DNSKEY RRset. The key roll is now
+complete though provider A SHOULD monitor the propagation of its DNSKEY
+RRset and ask provider B to report on propagation as well.
+
+### KSK rolls
+
+RFC 6781 defines two KSK rolls: the Double-Signature KSK roll and the
+Double-DS KSK roll. The Double-Signature KSK roll is easiest and we will
+look that first.
+
+We assume provider A wants to roll to KSK_An. First provider A adds
+KSK_An to its DNSKEY RRset. Provider A monitors the propagation of its
+DNSKEY RRset.
+
+When that has happened, provider A and B need to agree on new CDS/CDNSKEY
+RRsets. Provider A needs the new CDS/CDNSKEY RRset to contain or refer to
+KSK_An. When agreement has been reached, provider A publishes the CDS/CDNSKEY
+RRset and assumes that provider B will do so as well. 
+
+Provider A checks that the parent has a DS record for KSK_An and tries to 
+monitor to propagation of the DS RRset.
+
+When the DS RRset has fully propagated, provider A removes KSK_A from its
+DNSKEY RRset, and tries to reach agreement to remove the CDS/CDNSKEY RRset.
+
+The key roll is now complete but provider A SHOULD monitor the propagation
+of its new DNSKEY RRset.
+
+The Double-DS KSK roll starts with introducing the new KSK_An in the parent's
+DS RRset. Provider A and B need to agree on a new CDS/CDNSKEY RRsets that
+include both KSK_A and KSK_An. When agreement has been reached, provider A
+publishes the CDS/CDNSKEY RRset and assumes that provider B will do so as
+well.
+
+Provider A checks that the parent has a DS record for both KSK_A and KSK_An
+and tries to monitor to propagation of the DS RRset.
+
+When the DS RRset has fully propagated, provider A removes KSK_A from its
+DNSKEY RRset and adds KSK_An. Provider A monitors the propagation of the
+new DNSKEY RRset.
+
+When the DNSKEY RRset has fully propagated, provider A tries to agree with
+provider B on new CDS/CDNSKEY RRsets that include KSK_An (and no longer KSK_A).
+When agreement has been reached, provider A
+publishes the CDS/CDNSKEY RRset and assumes that provider B will do so as
+well.
+
+Provider A checks that the parent has a DS record for KSK_An. Then
+provider A tries to reach agreement with provider B to remove the CDS/CDNSKEY
+RRset.
+
+(Alternatively: both provider A and B monitor the DS RRset set the parent.
+When the DS RRset matches the agreed upon set of keys, each provider
+independently removes to CDS and CDNSKEY RRsets. Is there a risk?)
```

Comment:
I think this is preferred.

(Also, in practice many providers don't remove CDS/CDNSKEY RRsets - probably because they have less incentive to monitor in single-signer scenarios. But the point is they haven't been hurting.)